### PR TITLE
Services/UDS: Handle the rest of the connection sequence.

### PR DESCRIPTION
**Pull request by B3n30**
Follow up of #2920. This is the last step in the connection sequence. 
Next steps are sending packets, receiving packets, disconnect and destroy.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2963)

### CODE REVIEW

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 1]
```diff
@@ -143,18 +160,80 @@ void HandleAssociationResponseFrame(const Network::WifiPacket& packet) {
...
+            LOG_DEBUG(Service_NWM, "Connection sequence aborted, because connection status is %u",
+                      connection_status.status);
+            return;
+        }
+
+        auto node = DeserializeNodeInfoFromFrame(packet.data);
+
+        if (connection_status.max_nodes == connection_status.total_nodes) {
+            // Reject connection attempt
+            // TODO(B3N30): Figure out what packet is sent here
```
Comment (Subv) [inline src/core/hle/service/nwm/nwm_uds.cpp]: We should log something here

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 2]
```diff
@@ -143,18 +160,80 @@ void HandleAssociationResponseFrame(const Network::WifiPacket& packet) {
...
+        for (size_t index = 0; index < logoff.connected_nodes; ++index) {
+            connection_status.node_bitmask |= 1 << index;
+            connection_status.changed_nodes |= 1 << index;
+            connection_status.nodes[index] = logoff.nodes[index].network_node_id;
+
+            node_info.emplace_back(DeserializeNodeInfo(logoff.nodes[index]));
+        }
+
+        // We're now connected, signal the application
+        connection_status.status = static_cast<u32>(NetworkStatus::ConnectedAsClient);
```
Comment (Subv) [inline src/core/hle/service/nwm/nwm_uds.cpp]: Shouldn't you also signal the event here?

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 3]
```diff
@@ -143,18 +160,80 @@ void HandleAssociationResponseFrame(const Network::WifiPacket& packet) {
...
+        WifiPacket eapol_logoff;
+        eapol_logoff.channel = network_channel;
+        eapol_logoff.data =
+            GenerateEAPoLLogoffFrame(packet.transmitter_address, node.network_node_id, node_info,
+                                     network_info.max_nodes, network_info.total_nodes);
+        // TODO(Subv): Encrypt the packet.
+        eapol_logoff.destination_address = packet.transmitter_address;
+        eapol_logoff.type = WifiPacket::PacketType::Data;
+
+        SendPacket(eapol_logoff);
```
Comment (Subv) [inline src/core/hle/service/nwm/nwm_uds.cpp]: Add a todo saying that we should broadcast the updated node list when a new client connects .

The 3ds does this presumably to support spectators

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 4]
```diff
@@ -143,18 +160,80 @@ void HandleAssociationResponseFrame(const Network::WifiPacket& packet) {
...
+
+        CoreTiming::ScheduleEvent_Threadsafe_Immediate(connection_status_changed_event);
```
Comment (Subv) [inline src/core/hle/service/nwm/nwm_uds.cpp]: You should lock the hle mutex and signal the event directly

File: src/core/hle/service/nwm/uds_data.cpp [hunk 5]
```diff
@@ -295,5 +295,80 @@ std::vector<u8> GenerateEAPoLStartFrame(u16 association_id, const NodeInfo& node
...
+NodeInfo DeserializeNodeInfoFromFrame(const std::vector<u8>& frame) {
+    EAPoLStartPacket eapol_start;
+
+    // Skip the LLC header
+    std::memcpy(&eapol_start, frame.data() + sizeof(LLCHeader), sizeof(eapol_start));
+
+    NodeInfo node{};
+    node.friend_code_seed = eapol_start.node.friend_code_seed;
+
+    for (int i = 0; i < node.username.size(); ++i)
```
Comment (lioncash) [inline src/core/hle/service/nwm/uds_data.cpp]: FWIW, these can use `std::copy`

```cpp
std::copy(eapol_start.node.username.begin(),
          eapol_start.node.username.end(), 
          node.username.begin());
```

File: src/core/hle/service/nwm/uds_data.cpp [hunk 6]
```diff
@@ -295,5 +295,80 @@ std::vector<u8> GenerateEAPoLStartFrame(u16 association_id, const NodeInfo& node
...
+
+    return node;
+}
+
+NodeInfo DeserializeNodeInfo(const EAPoLNodeInfo& node) {
+    NodeInfo node_info{};
+    node_info.friend_code_seed = node.friend_code_seed;
+    node_info.network_node_id = node.network_node_id;
+
+    for (int i = 0; i < node.username.size(); ++i)
```
Comment (lioncash) [inline src/core/hle/service/nwm/uds_data.cpp]:
```cpp
std::copy(node.username.begin(), node.username.end(), node_info.username.begin());
```

File: src/core/hle/service/nwm/uds_data.cpp [hunk 7]
```diff
@@ -295,5 +295,80 @@ std::vector<u8> GenerateEAPoLStartFrame(u16 association_id, const NodeInfo& node
...
+    eapol_logoff.max_nodes = max_nodes;
+
+    for (size_t index = 0; index < total_nodes; ++index) {
+        const auto& node_info = nodes[index];
+        auto& node = eapol_logoff.nodes[index];
+
+        node.friend_code_seed = node_info.friend_code_seed;
+        node.network_node_id = node_info.network_node_id;
+
+        for (int i = 0; i < node.username.size(); ++i)
```
Comment (lioncash) [inline src/core/hle/service/nwm/uds_data.cpp]:
```cpp
std::copy(node_info.username.begin(), node_info.username.end(), node.username.end());
```

Comment (B3n30) [reply src/core/hle/service/nwm/uds_data.cpp]: `std::copy(node_info.username.begin(), node_info.username.end(), node.username.begin());`

Comment (lioncash) [reply src/core/hle/service/nwm/uds_data.cpp]: close enough :^)

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 8]
```diff
@@ -143,18 +158,88 @@ void HandleAssociationResponseFrame(const Network::WifiPacket& packet) {
...
+            connection_status.node_bitmask |= 1 << index;
+            connection_status.changed_nodes |= 1 << index;
+            connection_status.nodes[index] = logoff.nodes[index].network_node_id;
+
+            node_info.emplace_back(DeserializeNodeInfo(logoff.nodes[index]));
+        }
+
+        // We're now connected, signal the application
+        connection_status.status = static_cast<u32>(NetworkStatus::ConnectedAsClient);
+        // Some games require ConnectToNetwork beeing blocking, for now it's not
```
Comment (Subv) [inline src/core/hle/service/nwm/nwm_uds.cpp]: being

Comment (BreadFish64) [reply src/core/hle/service/nwm/nwm_uds.cpp]: even better: to be blocked

Comment (Subv) [reply src/core/hle/service/nwm/nwm_uds.cpp]: `Some games require ConnectToNetwork to block`

**Comment by Subv** [comment]
Please squash the commits so we can merge

**Pull request merged**